Documentation/design-docs/native-hosting-initialize.md

```diff
@@ -0,0 +1,213 @@
+# Native hosting with explicit initialize
+
+## Goals
+* All hosts should be able to use the new API (whether they will is a separate question as the old API has to be kept for backward compat reasons)
+* Hide implementation details as much as possible
+  * Make the API generally easier to use/understand
+  * Give the implementation more freedom
+  * Allow future improvements without breaking the API
+  * Consider explicitly documenting types of behaviors which nobody should take dependency on (specifically failure scenarios)
+* Extensible
+
+## New scenarios
+The API should allow these scenarios:
+* Runtime properties
+  * Specify additional runtime properties from the native host
+  * Implement conflict resolution for runtime properties
+  * Inspect calculated runtime properties (the ones calculated by `hostfxr`/`hostpolicy`)
+
+It should be possible to ship with only some of these supported, then enable more scenarios later on.
+
+## New `hostfxr` APIs
+
+Terminology:
+* "native host" - the code which uses the proposed APIs. Typically a native (C++) application, but can really be any "non .NET Core" code.
+* "host context" - state which `hostfxr` creates and maintains and represents a logical operation on the hosting components.
+* "hosting components" - shorthand for .NET Core hosting components. Typically refers to `hostfxr` and `hostpolicy`. Sometimes also referred to simply as "host".
+
+### Initialize host context
+
+All the "initialize" functions will
+* Process the `.runtimeconfig.json`
+* Resolve framework references and find actual frameworks
+* Find the root framework (`Microsoft.NETCore.App`) and load the `hostpolicy` from it
+* The `hostpolicy` will then process all relevant `.deps.json` files and produce the TPA, native search paths and other artifacts.
```

Comment:
There are so many acronyms in the hosting world. Can we define "TPA" and "TFM" or at least provide a link to a doc with the definitions.

Reply:
Removed the acronym...